\nonstopmode{}
\documentclass[a4paper]{book}
\usepackage[times,inconsolata,hyper]{Rd}
\usepackage{makeidx}
\usepackage[utf8]{inputenc} % @SET ENCODING@
% \usepackage{graphicx} % @USE GRAPHICX@
\makeindex{}
\begin{document}
\chapter*{}
\begin{center}
{\textbf{\huge Package `optima'}}
\par\bigskip{\large \today}
\end{center}
\inputencoding{utf8}
\ifthenelse{\boolean{Rd@use@hyper}}{\hypersetup{pdftitle = {optima: What the Package Does (One Line, Title Case)}}}{}
\ifthenelse{\boolean{Rd@use@hyper}}{\hypersetup{pdfauthor = {First Last}}}{}
\begin{description}
\raggedright{}
\item[Title]\AsIs{What the Package Does (One Line, Title Case)}
\item[Version]\AsIs{0.0.0.9000}
\item[Description]\AsIs{What the package does (one paragraph).}
\item[License]\AsIs{`use_mit_license()`, `use_gpl3_license()` or friends to pick a
license}
\item[Encoding]\AsIs{UTF-8}
\item[Roxygen]\AsIs{list(markdown = TRUE)}
\item[RoxygenNote]\AsIs{7.2.3}
\end{description}
\Rdcontents{\R{} topics documented:}
\inputencoding{utf8}
\HeaderA{annotateVariant}{Variant annotation}{annotateVariant}
\keyword{variant}{annotateVariant}
%
\begin{Description}\relax
This function takes variant names as input and
returns annotation for annotation table for all variant IDs in a data frame.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
annotateVariant(variant.names)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{variant.names}] input variant IDs, can be a vector.
\end{ldescription}
\end{Arguments}
%
\begin{Value}
a data frame with annotation for all input variant IDs.
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
annotateVariant(variants_id)

\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{calculatePloidy}{CNV ploidy calculation function}{calculatePloidy}
\keyword{being}{calculatePloidy}
\keyword{optima.obj}{calculatePloidy}
\keyword{ploidy.mtx}{calculatePloidy}
\keyword{updated}{calculatePloidy}
\keyword{with}{calculatePloidy}
%
\begin{Description}\relax
The function used normalized CNV matrix to calculate the ploidy for each CNV locus.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
calculatePloidy(optima.obj, diploid.cell)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{optima.obj}] optima object.

\item[\code{diploid.cell}] the cell type that should be considered as diploid cell
\end{ldescription}
\end{Arguments}
%
\begin{Value}
optima object with normalized CNV.
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
calculatePloidy(optima.obj)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{drawHeatmap}{Heatmap function}{drawHeatmap}
%
\begin{Description}\relax
This function creates a heatmap using matrix data, such matrix
data can be DNA or protein.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
drawHeatmap(optima.obj, omic.type)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{optima.obj}] optima object.

\item[\code{omic.type}] Type of data for heat map. Potential values "dna" and "protein".
\end{ldescription}
\end{Arguments}
%
\begin{Value}
Heat map visualization.
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
drawHeatMap()
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{filterVariant}{DNA Variant filter function}{filterVariant}
\keyword{DNA}{filterVariant}
\keyword{filter}{filterVariant}
%
\begin{Description}\relax
This function uses multiple matrices imported from the h5 file to conduct quality filtering.
This includes the sequencing depth matrix, genotype matrix, variant allele frequency matrix, genotype quality matrix
The function returns an optima object that has been filtered with variant/cells.
In addition, the returned optima object's variant.filter label is changed to "filtered".
This function is usually applied before protein and CNV analysis.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
filterVariant(
  optima.obj,
  min.dp = 10,
  min.gq = 30,
  vaf.ref = 5,
  vaf.hom = 95,
  vaf.het = 35,
  min.cell.pt = 50,
  min.mut.cell.pt = 1
)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{optima.obj}] an optima object with raw data unfiltered.

\item[\code{min.dp}] minimum depth, defaults to 10

\item[\code{min.gq}] minimum genotype quality, defaults to 30

\item[\code{vaf.ref}] If reference call vaf (GT=0) is larger than vaf.ref, then value in genotype call matrix is converted to GT=3

\item[\code{vaf.hom}] If homozygous call vaf (GT=2) is smaller than vaf.hom, then value in genotype call matrix is converted to GT=3

\item[\code{vaf.het}] If heterozygous call vaf (GT=1) is smaller than vaf.ref, then value in genotype call matrix is converted to GT=3

\item[\code{min.cell.pt}] minimum threshold for cell percentage that has valid variant call (GT = 0, 1 or 2) after applying the filter.

\item[\code{min.mut.cell.pt}] minimum threshold for cell percentage that has mutated genotype (GT = 1 or 2) after applying the filter.
\end{ldescription}
\end{Arguments}
%
\begin{Value}
an optima object, The DNA data in the object is filtered, the variant.filter label is "filtered".
Meanwhile, the protein matrix and CNV matrix is also updated so that only cells withstand DNA variant filter are kept.
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
filterVariant(optima.obj)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{findSignature}{Identify signature protein function}{findSignature}
\keyword{cell.type}{findSignature}
\keyword{optima.obj,}{findSignature}
%
\begin{Description}\relax
This function compares protein levels for a input cell type against all other
cells using t test. This function returns a data frame ranked by
FDR adjusted p-value.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
findSignature(optima.obj, cell.type)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{optima.obj}] optima object.

\item[\code{cell.type}] Input cell type to compare protein level to all other cell types.
\end{ldescription}
\end{Arguments}
%
\begin{Value}
Data frame of all proteins p-values comparing protein levels of input
cell type to all other cell types.
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
findSignature(optima.obj, cell.type)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{getClones}{Clustering variant function}{getClones}
%
\begin{Description}\relax
This function identifies cell clones based on DNA variant data.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
getClones(optima.obj, eps = 1, minPts = 100, plot = FALSE)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{optima.obj}] optima object.

\item[\code{eps}] size/radius of the epsilon neighborhood.
This argument will passed to dbscan function.

\item[\code{minPts}] number of minimum points required in the eps neighborhood
for core points, including the point itself.
This argument will passed to dbscan function.

\item[\code{plot}] if True, a UMAP plot will be generated based on the dimension reduction
result from variant matrix.
Default is FALSE.
\end{ldescription}
\end{Arguments}
%
\begin{Value}
Data frame with annotation for all input variant IDs.
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
getClones(my.obj)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{getInfo}{Single variant ID annotation function}{getInfo}
\keyword{variant}{getInfo}
%
\begin{Description}\relax
Returns annotation from one variant ID. This function is not visable to users
\end{Description}
%
\begin{Usage}
\begin{verbatim}
getInfo(variant)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{variant}] variant name in a string
\end{ldescription}
\end{Arguments}
%
\begin{Value}
Annotation information for one specific variant ID.
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
getInfo(variant_id)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{normalizeCNV}{CNV normalization function}{normalizeCNV}
\keyword{optima.obj}{normalizeCNV}
%
\begin{Description}\relax
The function normalizes the CNV matrix to correct for column wise and row wise variation and
updates the optima object amp.normalize.method from "unnormalized" to "normalized".
\end{Description}
%
\begin{Usage}
\begin{verbatim}
normalizeCNV(optima.obj)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{optima.obj}] optima object.
\end{ldescription}
\end{Arguments}
%
\begin{Value}
optima object with normalized CNV. the amp.normalize.method will be updated to "normalized"
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
normalizeCNV(optima.obj)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{normalizeProtein}{Protein matrix normalization}{normalizeProtein}
\keyword{optima.obj}{normalizeProtein}
%
\begin{Description}\relax
The function normalizes protein matrix within an optima object using
CLR method.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
normalizeProtein(optima.object)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{optima.obj}] optima object
\end{ldescription}
\end{Arguments}
%
\begin{Value}
an optima object with protein matrix being normalized and
protein.normalize.method label updated to "normalized"
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
normalizeProtein(optima.object)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{optima-class}{optima object}{optima.Rdash.class}
\aliasA{optima}{optima-class}{optima}
%
\begin{Description}\relax
An optima object contains DNA, protein and CNV for Tapestri platform
single cell sequencing data.
\end{Description}
%
\begin{Value}
Object containing DNA, protein and CNV single cell sequencing data.
\end{Value}
%
\begin{Section}{Slots}

\begin{description}

\item[\code{meta.data}] user defined metadata can be kept with the object

\item[\code{cell.ids}] a vector of cell IDs/barcodes from Tapestri. This
vector should contain unique IDs

\item[\code{cell.labels}] a vector that is used to store the cell type information
for each cell

\item[\code{variants}] a vector of variant IDs

\item[\code{variant.filter}] a string that keeps track of if optima object is being QC filtered
on its variant matrix

\item[\code{vaf.mtx}] variant matrix

\item[\code{gt.mtx}] genotype matrix

\item[\code{dp.mtx}] sequencing depth matrix

\item[\code{gq.mtx}] genotype quality

\item[\code{amps}] a vector of CNV locus

\item[\code{amp.normalized.method}] a string that keeps track of if optima object is being normalized
on its CNV matrix

\item[\code{amp.mtx}] CNV matrix

\item[\code{ploidy.mtx}] ploidy matrix

\item[\code{proteins}] a vector of surface protein id

\item[\code{protein.normalize.method}] a string that keeps track of if optima object is being normalized
on its protein matrix

\item[\code{protein.mtx}] protein matrix

\end{description}
\end{Section}
%
\begin{Examples}
\begin{ExampleCode}
setClass()
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{plotPloidy}{Ploidy scatter plot function}{plotPloidy}
\keyword{optima.obj}{plotPloidy}
%
\begin{Description}\relax
For a specified cell type, the function creates a scatter plot indicating
ploidy for different CNV loci.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
plotPloidy(optima.obj, cell.type)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{optima.obj}] optima object.

\item[\code{cell.type}] String that indicates which cell type.
\end{ldescription}
\end{Arguments}
%
\begin{Value}
optima object with normalized CNV.
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
plotPloidy()
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{readHdf5}{H5 file to optima object function}{readHdf5}
\keyword{directory}{readHdf5}
%
\begin{Description}\relax
This function read in a h5 file and return one optima object. The
h5 file can be found in the Tapestri pipeline software output. The .h5
file contains all necessary data needed for single cell DNA and protein
analysis.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
readHdf5(directory)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{directory}] Directory for the input h5 file .
\end{ldescription}
\end{Arguments}
%
\begin{Value}
optima object
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
readHdf5("path/to/my/file.h5")
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{reduceDim}{Dimension reduction function.}{reduceDim}
%
\begin{Description}\relax
This function reduces dimensions for a data matrix, such data matrix
can be protein or DNA matrix in an optima object.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
reduceDim(input.mtx)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{input.mtx}] Input optima object.
\end{ldescription}
\end{Arguments}
%
\begin{Value}
List containing PCA result and umap result derived from first 5 PCs
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
reduceDim(example.matrix)
\end{ExampleCode}
\end{Examples}
\printindex{}
\end{document}
